# THIS IS A CONTENT BASED MOVIE RECOMMENDER SYSTEM

In [2]:
# from google.colab import drive
# drive.mount('/content/drive')

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
mov=pd.read_csv('tmdb_5000_movies.csv')
cred=pd.read_csv('tmdb_5000_credits.csv')

In [5]:
mov.head(5)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   
3  [{"iso_3166_1": "US", "name": "United States o...   2012-07-16  1084939099   
4  [{"iso_3166_1": "US", "name": "United States o...   2012-03-07   284139100   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3    165.0           [{"iso_639_1": "en", "name": "English"}]  Released   
4    132.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124

In [6]:
cred.head(5)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [7]:
mov.shape

(4803, 20)

In [8]:
cred.shape

(4803, 4)

### MERGING BOTH THE DATAFRAMES INTO ONE DATAFRAME

In [10]:
all_df=mov.merge(cred,on='title')

In [11]:
all_df.shape

(4809, 23)

In [12]:
all_df.head(5)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  ... runtime  \
0  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...  ...   169.0   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...  ...   148.0   
3  [{"name": "Legendary Pictures", "id": 923}, {"...  ...   165.0   
4        [{"name": "Walt Disney Pictures", "id": 2}]  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count movie_id  \
0                                    Avatar          7.2      11800    19995   
1  Pirates of the Caribbean: At World's End          6.9       4500      285   
2                                   Spectre          6.3       4466   206647   
3                     The Dark Knight Rises          7.6       9106    49026   
4                               John Carter          6.1       2124    49529   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

In [13]:
all_df.isnull().sum()

budget                     0
genres                     0
homepage                3096
id                         0
keywords                   0
original_language          0
original_title             0
overview                   3
popularity                 0
production_companies       0
production_countries       0
release_date               1
revenue                    0
runtime                    2
spoken_languages           0
status                     0
tagline                  844
title                      0
vote_average               0
vote_count                 0
movie_id                   0
cast                       0
crew                       0
dtype: int64

## As I'm building a Content-based Movie Recommender System, I am including only columns or variables, which have categorical data.

In [15]:
all_df=all_df[['id','title','genres','overview','keywords','cast','crew']]
all_df.head(3)

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...

In [16]:
all_df.shape

(4809, 7)

In [17]:
all_df.isnull().sum()

id          0
title       0
genres      0
overview    3
keywords    0
cast        0
crew        0
dtype: int64

In [18]:
np.round(all_df.isnull().mean()*100,2) # it gives the percentage of null values

id          0.00
title       0.00
genres      0.00
overview    0.06
keywords    0.00
cast        0.00
crew        0.00
dtype: float64

In [19]:
all_df.dropna(inplace=True)

In [20]:
np.round(all_df.isnull().mean()*100,2)

id          0.0
title       0.0
genres      0.0
overview    0.0
keywords    0.0
cast        0.0
crew        0.0
dtype: float64

In [21]:
all_df.duplicated().sum() # This implies there are no duplicated records or rows.

0

In [22]:
all_df['genres'][0]

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

### AS all_df['genres'] is a STRING in which there is a LIST OF DICTIONARIES WHICH CONTAINS GENRES, WE NEED TO CONVERT IT INTO A LIST FIRST

In [24]:
import ast

### Using "ast" MODULE FOR COVERTING STRING (in which it contains lists) TO LISTS, BY REMOVING THE STRING

In [26]:
ast.literal_eval(all_df['genres'][0]) # IT WORKS LIKE THIS

[{'id': 28, 'name': 'Action'},
 {'id': 12, 'name': 'Adventure'},
 {'id': 14, 'name': 'Fantasy'},
 {'id': 878, 'name': 'Science Fiction'}]

In [27]:
# Function for converting strings (in which it contains lists) into lists
def release(obj):
    l=[]
    for i in ast.literal_eval(obj):
       l.append(i['name'])
    return l

In [28]:
all_df['genres']=all_df['genres'].apply(release) # applting function to all_df['genres']

In [29]:
all_df.sample(3)

id                                      title  \
2959   1590                          Copying Beethoven   
1084   9381                               Babylon A.D.   
4373  22597  The Broken Hearts Club: A Romantic Comedy   

                                              genres  \
2959                                         [Drama]   
1084  [Action, Adventure, Science Fiction, Thriller]   
4373                        [Comedy, Drama, Romance]   

                                               overview  \
2959  A fictionalised exploration of Beethoven's lif...   
1084  In Babylon A.D Vin Diesel stars as a veteran-t...   
4373  A group of friends search for fun, love and ul...   

                                               keywords  \
2959  [{"id": 766, "name": "beethoven"}, {"id": 1870...   
1084  [{"id": 339, "name": "submarine"}, {"id": 1009...   
4373  [{"id": 237, "name": "gay"}, {"id": 1253, "nam...   

                                                   cast  \
2959  [{"cast_id": 4, "character": "Ludwig van Beeth...   
1084  [{"cast_id": 14, "character": "Toorop", "credi...   
4373  [{"cast_id": 1, "character": "Patrick", "credi...   

                                                   crew  
2959  [{"credit_id": "52fe4302c3a36847f80337af", "de...  
1084  [{"credit_id": "52fe44f0c3a36847f80b2e2f", "de...  
4373  [{"credit_id": "53dd58e9c3a3686c5a00021d", "de...

In [30]:
all_df['keywords'][0] # it is in same format as in "genres" variable

'[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]'

In [31]:
all_df['keywords']=all_df['keywords'].apply(release) # So, Applying "release" function to all_df['keywords']
all_df.sample(3)

id                   title                          genres  \
734   10199  The Tale of Despereaux  [Adventure, Animation, Family]   
1078   9486          Johnny English     [Adventure, Action, Comedy]   
3994  21461    Lage Raho Munna Bhai        [Comedy, Drama, Romance]   

                                               overview  \
734   Once upon a time... in the far away kingdom of...   
1078  Rowan plays the eponymous lead character in a ...   
3994  A hilarious underworld gangster known as Munna...   

                                               keywords  \
734   [loyalty, totalitarian regime, mouse, forgiven...   
1078  [spy, hero, queen, intelligence, coronation, f...   
3994      [historical figure, comedy, bollywood, india]   

                                                   cast  \
734   [{"cast_id": 5, "character": "Despereaux (voic...   
1078  [{"cast_id": 1, "character": "Johnny English",...   
3994  [{"cast_id": 9, "character": "Murli Prasad Sha...   

                                                   crew  
734   [{"credit_id": "58bad317c3a368664d023ea1", "de...  
1078  [{"credit_id": "52fe44fdc3a36847f80b6459", "de...  
3994  [{"credit_id": "52fe441ac3a368484e00f9c3", "de...

In [32]:
all_df['cast'][0]

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

<h3>As we have enormous number of cast members in every record, I am just going with top three(3) casts i.e.</h3>
<p>-> {"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}</p>
<p>-> {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}</p>
<p>-> {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}</p>

In [34]:
# Function for converting strings(in which it contains lists) into lists
# (This function also selects top 3[three] cast names in the movie.)
def releaseTop3(obj):
    l=[]
    c=0
    for i in ast.literal_eval(obj):
        if c<=2:
            l.append(i['name'])
            c+=1
        else:
            break
    return l

In [35]:
all_df['cast']=all_df['cast'].apply(releaseTop3)

In [36]:
all_df.sample(3)

id             title                       genres  \
3871     598       City of God               [Drama, Crime]   
1839    2085    Romeo Must Die    [Action, Crime, Thriller]   
4446  127867  The Brass Teapot  [Comedy, Fantasy, Thriller]   

                                               overview  \
3871  Cidade de Deus is a shantytown that started du...   
1839  Two warring gang families (one African-America...   
4446  When a couple discovers that a brass teapot ma...   

                                               keywords  \
3871  [male nudity, street gang, brazilian, photogra...   
1839  [martial arts, hip-hop, oakland, asian, asian ...   
4446  [fantasy, based on comic book, money, magical ...   

                                                   cast  \
3871  [Alexandre Rodrigues, Leandro Firmino, Phellip...   
1839               [Jet Li, Aaliyah, Isaiah Washington]   
4446     [Juno Temple, Michael Angarano, Alexis Bledel]   

                                                   crew  
3871  [{"credit_id": "52fe425ac3a36847f8017a9f", "de...  
1839  [{"credit_id": "52fe4333c3a36847f804213d", "de...  
4446  [{"credit_id": "52fe4b0bc3a368484e174b71", "de...

In [37]:
all_df['crew'][235]

'[{"credit_id": "52fe4354c3a36847f804c0b1", "department": "Camera", "gender": 2, "id": 997, "job": "Director of Photography", "name": "Thierry Arbogast"}, {"credit_id": "52fe4354c3a36847f804c0b7", "department": "Production", "gender": 0, "id": 2424, "job": "Casting", "name": "Pierre-Jacques B\\u00e9nichou"}, {"credit_id": "52fe4354c3a36847f804c0e9", "department": "Production", "gender": 2, "id": 11204, "job": "Executive Producer", "name": "Pierre Grunstein"}, {"credit_id": "52fe4354c3a36847f804c0bd", "department": "Art", "gender": 0, "id": 13839, "job": "Set Designer", "name": "Aline Bonetto"}, {"credit_id": "52fe4354c3a36847f804c0f5", "department": "Editing", "gender": 2, "id": 19379, "job": "Editor", "name": "Yannick Kergoat"}, {"credit_id": "52fe4354c3a36847f804c087", "department": "Directing", "gender": 2, "id": 20674, "job": "Director", "name": "Thomas Langmann"}, {"credit_id": "52fe4354c3a36847f804c08d", "department": "Production", "gender": 2, "id": 20674, "job": "Producer", "na

### As we have enormous number of crew members in every record, I am just going with Director

In [39]:
# Function for converting strings(in which it contains lists) into lists and for Selecting movie director's name.
def release_director_name(obj):
    l=[]
    for i in ast.literal_eval(obj):
        if i['job']=='Director':
            l.append(i['name'])
    return l

In [40]:
all_df['crew'].apply(release_director_name)

0                                [James Cameron]
1                               [Gore Verbinski]
2                                   [Sam Mendes]
3                            [Christopher Nolan]
4                               [Andrew Stanton]
                          ...                   
4804                          [Robert Rodriguez]
4805                              [Edward Burns]
4806                               [Scott Smith]
4807                               [Daniel Hsia]
4808    [Brian Herzlinger, Jon Gunn, Brett Winn]
Name: crew, Length: 4806, dtype: object

In [41]:
all_df['crew'][4808]

'[{"credit_id": "58ce021b9251415a390165d9", "department": "Production", "gender": 2, "id": 6888, "job": "Executive Producer", "name": "Clark Peterson"}, {"credit_id": "58ce0232c3a36850e90157da", "department": "Production", "gender": 2, "id": 61051, "job": "Executive Producer", "name": "Andrew Reimer"}, {"credit_id": "52fe44e8c3a368484e03da8d", "department": "Directing", "gender": 2, "id": 85563, "job": "Director", "name": "Brian Herzlinger"}, {"credit_id": "52fe44e8c3a368484e03da87", "department": "Directing", "gender": 2, "id": 94471, "job": "Director", "name": "Jon Gunn"}, {"credit_id": "52fe44e8c3a368484e03da97", "department": "Directing", "gender": 0, "id": 997560, "job": "Director", "name": "Brett Winn"}]'

In [42]:
all_df['Director_name']=all_df['crew'].apply(release_director_name)

In [43]:
all_df.head(2)

id                                     title  \
0  19995                                    Avatar   
1    285  Pirates of the Caribbean: At World's End   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]   

                                                crew     Director_name  
0  [{"credit_id": "52fe48009251416c750aca23", "de...   [James Cameron]  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  [Gore Verbinski]

In [44]:
all_df.drop('crew',axis=1,inplace=True)

In [45]:
all_df.sample(5)

id                           title  \
3664   12486           Farce of the Penguins   
4308     429  The Good, the Bad and the Ugly   
2673    2604      Born on the Fourth of July   
1419   94348                      Alex Cross   
3123  245846                    Strangerland   

                                  genres  \
3664               [Comedy, Documentary]   
4308                           [Western]   
2673                        [Drama, War]   
1419  [Action, Thriller, Crime, Mystery]   
3123                   [Drama, Thriller]   

                                               overview  \
3664  In this spoof of "March of the Penguins," natu...   
4308  While the Civil War rages between the Union an...   
2673  The biography of Ron Kovic. Paralyzed in the V...   
1419  After Washington DC detective Alex Cross is to...   
3123  Newly arrived to a remote desert town, Catheri...   

                                               keywords  \
3664                       [penguin, balzen, antarctic]   
4308  [bounty hunter, refugee, gold, anti hero, gall...   
2673  [vietnam veteran, post traumatic stress  disor...   
1419                                      [ex military]   
3123  [australia, missing child, outback, alcoholic ...   

                                                   cast   Director_name  
3664  [Samuel L. Jackson, Jim Belushi, Whoopi Goldberg]     [Bob Saget]  
4308       [Eli Wallach, Clint Eastwood, Lee Van Cleef]  [Sergio Leone]  
2673      [Tom Cruise, Raymond J. Barry, Caroline Kava]  [Oliver Stone]  
1419           [Tyler Perry, Rachel Nichols, Jean Reno]     [Rob Cohen]  
3123      [Nicole Kidman, Joseph Fiennes, Hugo Weaving]   [Kim Farrant]

#### NOW,  WE HAVE TO REMOVE ALL THE SPACE BETWEEN EVERY WORD IN EVERY VARIABLE OF TAGS TO AVOID CONFUSION FOR THE MODEL i.e.
#### => we have to convert "Vishal Raj" a name to "VishalRaj"
##### AS TAGS WILL ALWAYS HAVE NO SPACE B/W THEM

In [47]:
all_df['genres']=all_df['genres'].apply(lambda i:[w.replace(" ","") for w in i])
all_df['keywords']=all_df['keywords'].apply(lambda i:[w.replace(" ","") for w in i])
all_df['cast']=all_df['cast'].apply(lambda i:[w.replace(" ","") for w in i])
all_df['Director_name']=all_df['Director_name'].apply(lambda i:[w.replace(" ","") for w in i])

In [48]:
all_df.sample(7)

id                                title  \
577     395              AVP: Alien vs. Predator   
3395   1956                                Gerry   
2182  61012           Silent Hill: Revelation 3D   
3708  65749                                Tango   
3696  10822                         Savage Grace   
2609    115                     The Big Lebowski   
114     674  Harry Potter and the Goblet of Fire   

                                   genres  \
577   [Adventure, ScienceFiction, Action]   
3395          [Mystery, Drama, Adventure]   
2182          [Thriller, Horror, Mystery]   
3708            [Drama, Foreign, Romance]   
3696                              [Drama]   
2609                      [Comedy, Crime]   
114          [Adventure, Fantasy, Family]   

                                               overview  \
577   When scientists discover something in the Arct...   
3395  Two friends named Gerry become lost in the des...   
2182  Heather Mason and her father have been on the ...   
3708  A dangerous love affair inspires a director to...   
3696  This examination of a famous scandal from the ...   
2609  Jeffrey "The Dude" Lebowski, a Los Angeles sla...   
114   Harry starts his fourth year at Hogwarts, comp...   

                                               keywords  \
577   [savingtheworld, predator, laserpointer, space...   
3395  [desperation, highway, wilderness, gettinglost...   
2182  [femaleprotagonist, anotherdimension, basedonv...   
3708                     [dancer, tango, musical, love]   
3696  [mother, sex, adultery, nudity, scandal, son, ...   
2609  [whiterussian, dude, bowling, vietnamveteran, ...   
114   [magic, dyinganddeath, broom, sorcerer'sappren...   

                                                cast       Director_name  
577            [SanaaLathan, RaoulBova, EwenBremner]  [PaulW.S.Anderson]  
3395                       [CaseyAffleck, MattDamon]        [GusVanSant]  
2182       [AdelaideClemens, KitHarington, SeanBean]  [MichaelJ.Bassett]  
3708    [MiguelÁngelSolá, CeciliaNarova, MíaMaestro]       [CarlosSaura]  
3696  [JulianneMoore, StephenDillane, EddieRedmayne]          [TomKalin]  
2609       [JeffBridges, JohnGoodman, JulianneMoore]          [JoelCoen]  
114       [DanielRadcliffe, RupertGrint, EmmaWatson]        [MikeNewell]

In [49]:
all_df['overview']=all_df['overview'].apply(lambda v:v.split())
# Converting "Overview" variable into a list to combine with other columns as it isin string format
all_df.sample(7)

id                   title  \
3210   13483                   Awake   
312    22972              Green Zone   
682    12177           The Love Guru   
2437    7304          Running Scared   
444     4147       Road to Perdition   
4765  250902    The Image Revolution   
1459    9311  Smilla's Sense of Snow   

                                         genres  \
3210                 [Thriller, Crime, Mystery]   
312   [War, Action, Adventure, Drama, Thriller]   
682                           [Comedy, Romance]   
2437           [Action, Crime, Drama, Thriller]   
444                    [Thriller, Crime, Drama]   
4765                              [Documentary]   
1459  [Action, Crime, Drama, Mystery, Thriller]   

                                               overview  \
3210  [While, undergoing, heart, surgery,, a, man, e...   
312   [During, the, U.S.-led, occupation, of, Baghda...   
682   [Born, in, America, and, raised, in, an, India...   
2437  [After, a, drug-op, gone, bad,, Joey, Gazelle,...   
444   [Mike, Sullivan, works, as, a, hit, man, for, ...   
4765  [Twenty, years, ago,, seven, superstar, artist...   
1459  [Smilla, Jaspersen,, half, Danish,, half, Gree...   

                                               keywords  \
3210                                                 []   
312          [weaponofmassdestruction, baghdad, iraqi ]   
682   [sport, icehockey, guru, comedy, bollywood, in...   
2437  [icehockey, racism, pedophile, throatslitting,...   
444   [basedongraphicnovel, homework, shotinthechin,...   
4765  [comicbook, biography, comicbookartist, imagec...   
1459           [copenhagen, inuit, comet, arctic, snow]   

                                                  cast     Director_name  
3210  [HaydenChristensen, JessicaAlba, TerrenceHoward]      [JobyHarold]  
312           [MattDamon, GregKinnear, BrendanGleeson]  [PaulGreengrass]  
682         [MikeMyers, JessicaAlba, JustinTimberlake]   [MarcoSchnabel]  
2437          [PaulWalker, CameronBright, VeraFarmiga]     [WayneKramer]  
444      [TomHanks, TylerHoechlin, JenniferJasonLeigh]       [SamMendes]  
4765               [ToddMcFarlane, RobLiefeld, JimLee]   [PatrickMeaney]  
1459        [JuliaOrmond, GabrielByrne, RichardHarris]     [BilleAugust]

#### COMBINING EVERY COLUMN EXCEPT 'ID' AND 'TITLE' AS ALL_TAGS TO GET ALL THE TAGS RELATED TO A MOVIE IN A SINGLE COLUMN

In [51]:
all_df['all_Tags']=all_df['genres']+all_df['overview']+all_df['keywords']+all_df['cast']+all_df['Director_name']
all_df.drop(['genres','overview','keywords','cast','Director_name'],axis=1, inplace=True)

In [52]:
all_df.sample(5)

id                    title  \
1586   21972                Like Mike   
1954  109091            The Counselor   
2461   74643               The Artist   
2357   19840  I Love You, Beth Cooper   
1403   49022       Something Borrowed   

                                               all_Tags  
1586  [Comedy, Drama, Family, Fantasy, Calvin, and, ...  
1954  [Thriller, Crime, Drama, A, rich, and, success...  
2461  [Drama, Comedy, Romance, Hollywood,, 1927:, As...  
2357  [Comedy, Romance, Nerdy, teenager, Denis, Coov...  
1403  [Comedy, Drama, Romance, Though, Rachel, (Ginn...

In [53]:
all_df['all_Tags'][0]

['Action',
 'Adventure',
 'Fantasy',
 'ScienceFiction',
 'In',
 'the',
 '22nd',
 'century,',
 'a',
 'paraplegic',
 'Marine',
 'is',
 'dispatched',
 'to',
 'the',
 'moon',
 'Pandora',
 'on',
 'a',
 'unique',
 'mission,',
 'but',
 'becomes',
 'torn',
 'between',
 'following',
 'orders',
 'and',
 'protecting',
 'an',
 'alien',
 'civilization.',
 'cultureclash',
 'future',
 'spacewar',
 'spacecolony',
 'society',
 'spacetravel',
 'futuristic',
 'romance',
 'space',
 'alien',
 'tribe',
 'alienplanet',
 'cgi',
 'marine',
 'soldier',
 'battle',
 'loveaffair',
 'antiwar',
 'powerrelations',
 'mindandsoul',
 '3d',
 'SamWorthington',
 'ZoeSaldana',
 'SigourneyWeaver',
 'JamesCameron']

#### CONVERTING LIST TO STRING

In [55]:
all_df['all_Tags']=all_df['all_Tags'].apply(lambda x:" ".join(x))

In [56]:
all_df.sample(5)

id                title  \
3911     284        The Apartment   
4583   23827  Paranormal Activity   
4155  174311        All Is Bright   
849     3580           Changeling   
4749   85860               The FP   

                                               all_Tags  
3911  Comedy Drama Romance Bud Baxter is a minor cle...  
4583  Horror Mystery After a young, middle class cou...  
4155  Comedy Drama Two French Canadian ne’er-do-well...  
849   Crime Drama Mystery Christine Collins is overj...  
4749  Comedy Two rival gangs fight for control of Fr...

In [57]:
all_df['all_Tags'][5]

"Fantasy Action Adventure The seemingly invincible Spider-Man goes up against an all-new crop of villain – including the shape-shifting Sandman. While Spider-Man’s superpowers are altered by an alien organism, his alter ego, Peter Parker, deals with nemesis Eddie Brock and also gets caught up in a love triangle. dualidentity amnesia sandstorm loveofone'slife forgiveness spider wretch deathofafriend egomania sand narcism hostility marvelcomic sequel superhero revenge TobeyMaguire KirstenDunst JamesFranco SamRaimi"

In [58]:
all_df['all_Tags']=all_df['all_Tags'].apply(lambda x:x.lower())

In [59]:
all_df

id                                     title  \
0      19995                                    Avatar   
1        285  Pirates of the Caribbean: At World's End   
2     206647                                   Spectre   
3      49026                     The Dark Knight Rises   
4      49529                               John Carter   
...      ...                                       ...   
4804    9367                               El Mariachi   
4805   72766                                 Newlyweds   
4806  231617                 Signed, Sealed, Delivered   
4807  126186                          Shanghai Calling   
4808   25975                         My Date with Drew   

                                               all_Tags  
0     action adventure fantasy sciencefiction in the...  
1     adventure fantasy action captain barbossa, lon...  
2     action adventure crime a cryptic message from ...  
3     action crime drama thriller following the deat...  
4     action adventure sciencefiction john carter is...  
...                                                 ...  
4804  action crime thriller el mariachi just wants t...  
4805  comedy romance a newlywed couple's honeymoon i...  
4806  comedy drama romance tvmovie "signed, sealed, ...  
4807  when ambitious new york attorney sam is sent t...  
4808  documentary ever since the second grade when h...  

[4806 rows x 3 columns]

In [60]:
# Removing HTML tags in "all_Tags" variable
from bs4 import BeautifulSoup

In [61]:
text_html='<h1><p>I am Akshay Kumar</p>'
s=BeautifulSoup(text_html,'html.parser')
print(s.get_text())

I am Akshay Kumar


In [62]:
def remove_htmlTags(text):
    s=BeautifulSoup(text,'html.parser')
    return s.get_text()
all_df['all_Tags']=all_df['all_Tags'].apply(remove_htmlTags)
all_df.sample(6)

C:\Users\HOME\AppData\Local\Temp\ipykernel_17612\3508911121.py:2: MarkupResemblesLocatorWarning: The input looks more like a filename than markup. You may want to open this file and pass the filehandle into Beautiful Soup.
  s=BeautifulSoup(text,'html.parser')


id                              title  \
1845  175555                  A Madea Christmas   
1266    9544                         FearDotCom   
931    13836             Race to Witch Mountain   
4724   18632                  A Dog's Breakfast   
1230  188161  A Million Ways to Die in the West   
179   270946             Penguins of Madagascar   

                                               all_Tags  
1845  comedy drama madea dispenses her unique form o...  
1266  horror thriller with four corpses on his hands...  
931   adventure family fantasy sciencefiction thrill...  
4724  comedy if you've never been good at anything i...  
1230  comedy western as a cowardly farmer begins to ...  
179   family animation adventure comedy skipper, kow...

In [63]:
# Removing punctuations and numbers
import string
puncs_nums=string.punctuation+"0123456789’–"
puncs_nums

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~0123456789’–'

In [64]:
test=all_df.all_Tags.iloc[223]
test

"action sciencefiction after years of outrunning ruthless bounty hunters, escaped convict riddick suddenly finds himself caught between opposing forces in a fight for the future of the human race. now, waging incredible battles on fantastic and deadly worlds, this lone, reluctant hero will emerge as humanity's champion - and the last hope for a universe on the edge of annihilation. prison dystopia matteroflifeanddeath outerspace intergalactictravel vindiesel thandienewton karlurban davidtwohy"

In [65]:
def remove_puncs_nums(text):
    for i in puncs_nums:
        text=text.replace(i,"")
    return text

In [66]:
remove_puncs_nums(test)

'action sciencefiction after years of outrunning ruthless bounty hunters escaped convict riddick suddenly finds himself caught between opposing forces in a fight for the future of the human race now waging incredible battles on fantastic and deadly worlds this lone reluctant hero will emerge as humanitys champion  and the last hope for a universe on the edge of annihilation prison dystopia matteroflifeanddeath outerspace intergalactictravel vindiesel thandienewton karlurban davidtwohy'

In [67]:
all_df['all_Tags'].iloc[5]

"fantasy action adventure the seemingly invincible spider-man goes up against an all-new crop of villain – including the shape-shifting sandman. while spider-man’s superpowers are altered by an alien organism, his alter ego, peter parker, deals with nemesis eddie brock and also gets caught up in a love triangle. dualidentity amnesia sandstorm loveofone'slife forgiveness spider wretch deathofafriend egomania sand narcism hostility marvelcomic sequel superhero revenge tobeymaguire kirstendunst jamesfranco samraimi"

In [68]:
all_df['all_Tags']=all_df['all_Tags'].apply(remove_puncs_nums)
all_df.sample(4)

id                     title  \
3395   1956                     Gerry   
3924  15582             Teen Wolf Too   
2081  31005            Moonlight Mile   
1632  72207  The Five-Year Engagement   

                                               all_Tags  
3395  mystery drama adventure two friends named gerr...  
3924  comedy fantasy family although awkward college...  
2081  romance drama as he copes with the death of hi...  
1632  comedy exactly one year after tom meets violet...

In [69]:
all_df['all_Tags'].iloc[5]

'fantasy action adventure the seemingly invincible spiderman goes up against an allnew crop of villain  including the shapeshifting sandman while spidermans superpowers are altered by an alien organism his alter ego peter parker deals with nemesis eddie brock and also gets caught up in a love triangle dualidentity amnesia sandstorm loveofoneslife forgiveness spider wretch deathofafriend egomania sand narcism hostility marvelcomic sequel superhero revenge tobeymaguire kirstendunst jamesfranco samraimi'

In [70]:
import nltk
from nltk.stem import WordNetLemmatizer
nltk.download('wordnet')
wnl=WordNetLemmatizer()

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\HOME\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [71]:
test=all_df.all_Tags.iloc[39]
test

'adventure action sciencefiction sam flynn the techsavvy and daring son of kevin flynn investigates his fathers disappearance and is pulled into the grid with the help of a mysterious program named quorra sam quests to stop evil dictator clu from crossing into the real world artificialintelligence secretidentity computerprogram dystopia computer deception duel motorcycle neonlight autocracy garretthedlund jeffbridges oliviawilde josephkosinski'

In [72]:
lemmatized_test = [wnl.lemmatize(word, pos="v") for word in test.split()]

lemmatized_test_str = " ".join(lemmatized_test)

lemmatized_test_str

'adventure action sciencefiction sam flynn the techsavvy and dare son of kevin flynn investigate his father disappearance and be pull into the grid with the help of a mysterious program name quorra sam quest to stop evil dictator clu from cross into the real world artificialintelligence secretidentity computerprogram dystopia computer deception duel motorcycle neonlight autocracy garretthedlund jeffbridges oliviawilde josephkosinski'

In [73]:
def lemmatizer(text):
    lemmatized_test=[wnl.lemmatize(word, pos="v") for word in text.split()]
    lemmatized_test_str=" ".join(lemmatized_test)
    return lemmatized_test_str

In [74]:
all_df['all_Tags']=all_df['all_Tags'].apply(lemmatizer)
all_df.sample(4)

id                       title  \
2060  44113                       Stone   
2237   3509            A Scanner Darkly   
2562  17127  The Object of My Affection   
1590   6023             P.S. I Love You   

                                               all_Tags  
2060  drama parole officer jack mabry have only a fe...  
2237  animation sciencefiction thriller an undercove...  
2562  comedy drama romance a pregnant new york socia...  
1590  drama romance a young widow discover that her ...

In [75]:
all_df.shape

(4806, 3)

In [76]:
# Vector Representation

In [77]:
from sklearn.feature_extraction.text import CountVectorizer
cv=CountVectorizer(max_features=5000,stop_words='english')

In [78]:
all_vectors=cv.fit_transform(all_df['all_Tags']).toarray()
all_vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [79]:
len(cv.get_feature_names_out())

5000

In [197]:
# These are all the features used
features=[]
for i in cv.get_feature_names_out():
    features.append(i)

print("These are all the features used\n\n\n")
print(", ".join(features))

These are all the features used



aaron, aaroneckhart, aarontaylorjohnson, abandon, abduct, abigailbreslin, abilities, ability, able, aboard, abuse, abusive, academic, academy, accept, access, accident, accidental, accidentally, acclaim, accompany, accomplish, account, accountant, accuse, ace, achieve, act, action, actionhero, actionpacked, activist, activities, activity, actor, actors, actress, actual, actually, ad, adam, adammckay, adams, adamsandler, adamshankman, adapt, adaptation, add, addict, addiction, adjust, admire, admit, adolescence, adolescent, adopt, adoption, adrienbrody, adult, adultanimation, adultery, adulthood, adults, advance, advantage, adventure, advertise, advice, affair, affairs, affect, affection, affections, affluent, afghanistan, afraid, africa, african, africanamerican, aftercreditsstinger, afterlife, aftermath, age, agedifference, agency, agenda, agent, agents, aggressive, ago, agree, ahead, aid, aidanquinn, ail, aim, air, airplane, airplanecrash, airport, 

In [81]:
# Generating similarity scores using cosine_similarity for getting nearest vectors (i.e. recommended movies) 
# for given input (i.e. movie)

In [82]:
from sklearn.metrics.pairwise import cosine_similarity

In [83]:
cosine_similarity(all_vectors).shape

(4806, 4806)

In [84]:
all_similarities=cosine_similarity(all_vectors)

In [85]:
all_similarities[1]

array([0.08858079, 1.        , 0.06253054, ..., 0.02514977, 0.        ,
       0.        ])

In [86]:
# RETURNS FIRST FIVE NEAREST VECTORS INDICES WHEN GIVEN A INPUT VECTOR INDEX, hERE, INPUT VECTOR INDEX IS:1
sorted(list(enumerate(all_similarities[1])),reverse=True,key=lambda c:c[1])[1:6]

[(12, 0.4412487516463849),
 (199, 0.2988812888320308),
 (17, 0.27371875400769585),
 (216, 0.21552636243212991),
 (2131, 0.20526320231631418)]

In [87]:
def recommend(movie):
    movie_index=all_df[all_df['title']==movie].index[0]
    distances=all_similarities[movie_index]
    movies_list=sorted(list(enumerate(distances)),reverse=True,key=lambda c:c[1])[1:6]
    print('\nRecommended movies based on the input movie you entered:\n')
    for i in movies_list:
        print(all_df.iloc[i[0]].title)

In [88]:
all_movies_titles_available=all_df['title'].values
# for available_movie in sorted(all_movies_titles_available):
    # print(available_movie) # See movies in "all_movies_list.html"

In [89]:
try:
    movie_name=input('Please enter the movie you like from the above listed movies, so that we can recommend movies that are similar to the given movie:\n')
    recommend(movie_name)
except:
    print(f"\n{movie_name} movie is not available in this project.\nPlease provide another movie name for recommendations.")

Please enter the movie you like from the above listed movies, so that we can recommend movies that are similar to the given movie:
 Spider-Man



Recommended movies based on the input movie you entered:

Spider-Man 3
Spider-Man 2
The Amazing Spider-Man
Arachnophobia
Kick-Ass


In [90]:
len(all_movies_titles_available)

4806

In [91]:
dff=pd.DataFrame(all_movies_titles_available)
# dff.to_html('all_movies_list',index=False)

In [92]:
# import pickle

In [93]:
# pickle.dump(all_df,open('df.pkl','wb'))